### PR TITLE
PR: Add translator stub and example

### PR DESCRIPTION
**Pull request by goanpeca**
# Description

Related https://github.com/napari/napari/issues/2195

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by goanpeca** [comment]
@sofroniewn question. I am adding some tests that add a "language pack" package. Do I include that with napari or are _tests folder not normally bundled with the package?

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]: Why `Exception` not `ImportError`?

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]: There is plan to napari team grab all possible combinations?

Comment (goanpeca) [reply napari/utils/translations.py]: Hi @Czaki I do not follow what you mean :)

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py]: I will change :)

File: napari/utils/translations.py
Comment (Czaki) [reply napari/utils/translations.py]: I think about hostile takeover name on pypi, but backage on pypi could have a different name than installed package.

Comment (goanpeca) [reply napari/utils/translations.py]: True, will move to use entry points, that way is safer :) 

Good catch!

File: tox.ini
Comment (goanpeca) [inline tox.ini]: @tlambert03 this is what I tried but maybe I should do this differently?

Help :) !

**Comment by sofroniewn** [comment]
> @sofroniewn question. I am adding some tests that add a "language pack" package. Do I include that with napari or are _tests folder not normally bundled with the package?

Maybe @jni can weigh in on that, I can't remember where we landed. I'm also wondering if we should make the localization installation opt in (at least in terms of extra dependencies). We're already using pip install with extras for things like `pyside2` or `pyqt5` and have a `napari[all]` to catch everything). If this adds serious overhead and complexity and the extra dependencies are light weight maybe it's not needed, but just a thought.

**Comment by goanpeca** [comment]
[Babel ](http://babel.pocoo.org/en/latest/) is the extra dependency, which is pretty light in terms of other dependencies. (conda recipe on conda forge shows)

Screen Shot 2021-02-12 at 11 49 20 AM

We could also make babel a dependency of the `language-packs`, that way only when a language pack is installed, is when then Babel would be pulled.

This would be my suggestion that way no new deps are needed for napari.

**Comment by sofroniewn** [comment]
Thanks for that extra info - I'll defer to @jni here, but making a dependency of `language-packs` seems reasonable as without language packs you don't have a need for babel i guess. The only downside would be if our code becomes much harder to read because of it. Babel seems very light weight and widely used, so my concerns about it as a dependency are low.

File: tox.ini
Comment (tlambert03) [reply tox.ini]: probably should just put `napari/utils/_tests/napari-language-pack-es_CO` up in `deps`.

however, I'm trying to get translations to work locally on pytest and running into issues, here's what I tried, let me know what I should be doing:

```sh
pip install napari/utils/_tests/napari-language-pack-es_CO
...
Successfully installed napari-language-pack-es-CO-0.1.0
```

```python
In [1]: from napari.utils.translations import translator

In [2]: translator
Out[2]: napari.utils.translations.translator

In [3]: translator._set_locale("es_CO")

In [4]: trans = translator.load()

In [5]: trans
Out[5]: <napari.utils.translations.TranslationBundle at 0x7fea400986d0>

In [6]: trans._("MORE ABOUT NAPARI")
Out[6]: 'MORE ABOUT NAPARI'

In [7]: trans._locale
Out[7]: 'es_CO'

```

Comment (goanpeca) [reply tox.ini]: Hmm yes, I think I forgot to update the MANIFEST, the right way. I was using 

```
pip install -e napari/utils/_tests/napari-language-pack-es_CO
```


Or I went to the folder and did

```
pip install -e .
```

So the `*.mo` files were not found! Will fix. Thanks for checking!

Comment (tlambert03) [reply tox.ini]: ahh, cool.  OK, yeah, once that's done, I think moving the tox install up to the deps section (instead of the `commands_pre` section) _should_ do it ... but might need a trick for a local path?

Comment (goanpeca) [reply tox.ini]: Will test in a min.

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py]: 2 birds one stone...

Comment (goanpeca) [reply napari/utils/translations.py]: Hahaha yes! Need a rename 😆

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py]: Fixed

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py]: Fixed

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py]: Fixed

File: napari/utils/_tests/test_translations.py
Comment (goanpeca) [inline napari/utils/_tests/test_translations.py]: Ended up installing the package as part of the test @tlambert03 🤷🏼

**Comment by goanpeca** [comment]
Ok so I spend a lot of time debugging the installing of the package, then I decided to just install it as part of the actual tests and then trying to figure out why the test was failing. Turns out *.mo where on gitignore :-p ....

This is ready for review @sofroniewn, @tlambert03, @jni 

--- 

Update:

I removed babel imports as discussed and cleaned up a bit the tests.

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py]:
```suggestion
    cmd : list
```

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py]:
```suggestion
        A tuple in the form ``(result_dict, message)``.
```

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py]: jupyterlab ?

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py]:
```suggestion
        locale : str
```

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py:429]: :+1:

**Review by Carreau — APPROVED** [review]
LGTM !

File: napari/utils/_tests/test_translations.py
Comment (sofroniewn) [inline napari/utils/_tests/test_translations.py]: Our convention has been to use absolute imports in our tests, do you mind changing here and below

File: napari/utils/_tests/test_translations.py
Comment (sofroniewn) [reply napari/utils/_tests/test_translations.py]: this seems a little heavy. i'm not sure i have a better idea if the other installation methods didn't work. i wonder if there are any potential negative consequences if the installation fails, will the whole test run fail, or everything else will keep going as expected. I am ok with this though if we don't have other options. We can also just go with this for now and then maybe revisit sometime in the future, so no need to hold up the PR in my opinion

File: napari/utils/_tests/test_translations.py
Comment (sofroniewn) [inline napari/utils/_tests/test_translations.py]: Do we know anything about these fails? Does it seem just on CI, or also locally. We could add a skip that was specific to 3.7 ubuntu, we've done things like that in other places

File: napari/utils/translations.py
Comment (sofroniewn) [inline napari/utils/translations.py:26]: I think we generally have empty lines above `Parameters` and `Returns`

File: napari/utils/translations.py
Comment (sofroniewn) [inline napari/utils/translations.py:169]: Can you add a doc string here for the `domain` and `locale` Parameters in the init

**Review by sofroniewn — APPROVED** [review]
Ok, this looks like it's off to a great start. Most of my comments are fairly minor. Maybe @jni and @tlambert03 want to take a quick look at this, but I think it's good to go in now.

Once we've gotten this a bit further it would be great @goanpeca if you could write up a "localization and translation" guide, that could live here https://napari.org/docs/dev/guides/index.html when it is done with key information relevant to users, plugin developers, and napari maintainers.

File: napari/utils/_tests/test_translations.py
Comment (tlambert03) [reply napari/utils/_tests/test_translations.py]: yeah, lemme look into this before we merge.  will be busy for the weekend though.

File: napari/utils/translations.py
Comment (tlambert03) [inline napari/utils/translations.py]: this is returning a tuple, not dict

File: napari/utils/translations.py
Comment (tlambert03) [inline napari/utils/translations.py]: this will regress https://github.com/napari/napari/pull/2121 ... delay import (but preferably don't use at all, see comment below in `_get_installed_language_pack_locales`

File: napari/utils/translations.py
Comment (tlambert03) [inline napari/utils/translations.py]:
```suggestion
    if sys.version_info >= (3, 8):
        from importlib import metadata as importlib_metadata
    else:
        import importlib_metadata

    data = {}
    messages = []
    for dist in importlib_metadata.distributions():
        for ep in dist.entry_points:
            if ep.group == NAPARI_LANGUAGEPACK_ENTRY:
                try:
                    data[ep.name] = ep.value
                except Exception:
                    messages.append(traceback.format_exc())
```
we can avoid `pkg_resources` all together, which can add a lot of time to our imports depending on the user's environment (see https://github.com/napari/napari/pull/2121 and discussion in [this readme](https://github.com/takluyver/entrypoints)), by using the `importlib.metadata` from the standard library.  This is the entry point discovery mechanism that we use in napari-plugin-engine.

In fact, a lot of this entry-point discovery code is duplicating logic already contained in napari-plugin-engine...  I'll look into seeing how you might be able to use that

File: napari/utils/translations.py
Comment (tlambert03) [inline napari/utils/translations.py]: what's the motivation for the subprocess here?  (is it to avoid the slowness of pkg_resources?).  If it _is_ to achieve some level of concurrency, are there any potential race conditions in qt_main_window using `trans._`?

File: napari/utils/_tests/test_translations.py
Comment (tlambert03) [reply napari/utils/_tests/test_translations.py]: Yeah, now that I see how this is being used... I don't love it.  We've even run into issues (https://github.com/napari/napari/issues/1009) in the past by including complete "subpackages" with their own `setup.py` as test fixtures.  Tests also shouldn't have side-effects, and this test will permanently modify the user's development environment... which is less than desirable.

We have to do a similar sort of entry point [discovery test](https://github.com/napari/napari-plugin-engine/blob/master/testing/test_discovery.py) in napari-plugin-engine.  There I use the pytest `tmp_path` fixture to build a temporary plugin fixture that gets cleaned up at the end.  I can't remember off the top of my head how I get that discovered during tests, but it's possible and far preferable to literally running pip install in a subprocess).  If you're tired of messing around making these tests work, let me know and I can make a PR to this PR.

Also mentioned this elsewhere, but we can probably reuse a lot of the napari-plugin-engine mechanisms for translation plugin discovery here.

**Review by tlambert03 — CHANGES_REQUESTED** [review]
Love the functionality.  I do have some concerns about the implementation of discovering language plugins

- I definitely want to change the testing strategy to avoid `Popen(['pip', 'install'...])` with an actual setuptools package fixture. see comment below where I discuss how we test this in napari-plugin-engine.  I can help there if you want
- Also mentioned below, this duplicates a lot of the machinery that we already have in napari-plugin-engine for finding entry points.  I would be nice if we could try to keep our plugin discovery unified there, and avoid having two different ways to find plugins.  I can help there too if you'd like.
- this PR will revert https://github.com/napari/napari/pull/2121 and reintroduce the `pkg_resources` import as soon as a Viewer is created.  I commented inline on making that import lazy (and better still, avoiding `pkg_resources` altogether in favor of `importlib.metadata` from the standard library)
- why is the entry point discovery happening in a subprocess?
- not really this PR's issue... but phillip vendored a ton of .po files and translation stuff from humanize that will now be included in the dist with the changes to manifest in this PR.  I think we should follow this up with a PR that unvendors all of that stuff, just pulling in the two functions he's using (`naturalsize` and `intword`).

File: napari/utils/translations.py
Comment (tlambert03) [reply napari/utils/translations.py]: you can also use:
```python
from napari_plugin_engine.manager import iter_available_plugins

data = {k: v for (k, v, _) in iter_available_plugins(NAPARI_LANGUAGEPACK_ENTRY)}
```

File: napari/utils/translations.py
Comment (tlambert03) [reply napari/utils/translations.py:26]: is this function used anywhere besides the tests?  If not, can it just move to the tests file?

File: napari/utils/translations.py
Comment (tlambert03) [inline napari/utils/translations.py]: does this just fail silently if someone requests an unavailable locale?  Should we show a warning?

**Comment by tlambert03** [comment]
I created a PR to your branch with some recommended changes: https://github.com/goanpeca/napari/pull/1

File: napari/utils/_tests/test_translations.py
Comment (goanpeca) [reply napari/utils/_tests/test_translations.py]: Will do!

File: napari/utils/_tests/test_translations.py
Comment (goanpeca) [reply napari/utils/_tests/test_translations.py]: Thanks for the input,  the fixture can have a teardown and uninstall de test, so that would solve the side effect issue.

I do not have much experience with tox to be honest, but I did try different ideas to no avail.

>Also mentioned this elsewhere, but we can probably reuse a lot of the napari-plugin-engine mechanisms for translation plugin discovery here.

More on that on those comments.

File: napari/utils/_tests/test_translations.py
Comment (goanpeca) [reply napari/utils/_tests/test_translations.py]: >Do we know anything about these fails? Does it seem just on CI, or also locally. We could add a skip that was specific to 3.7 ubuntu, we've done things like that in other places

This was happenning only on linux and python3.7 on CI, I do not have a VM to test this myself locally. However it was a weird error since it was refering to missing methods on the gettext python module. But they should be there, so I was a bit puzzled.

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py:26]: This will be used to display the available languages in the preferences.

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py]: >what's the motivation for the subprocess here? (is it to avoid the slowness of pkg_resources?). If it is to achieve some level of concurrency, are there any potential race conditions in 

No, it is to be able to discover a langage-pack if it was installed after Napari is executed. Python entry points are "pre-discovered" when the python interpreter is instarted so:

If someone open napari, then installs a language pack in the ipyconsole, or another terminal and open preferences, it will not display this newly installed package.


>qt_main_window using trans._?

🤷🏼

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py]: Whhops!

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py:169]: Will do

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py]: We could, what is the normal way of rasining an error on napari?

Any convention?

**Comment by goanpeca** [comment]
>I created a PR to your branch with some recommended changes: goanpeca#1

Hi @tlambert03, thanks for the PR :)

The whole subprocess logic was used to be able to discover new language packs installed after napari started, which will be a normal way of installin languagepacks eventually.

Could you please confirm that this is the case?

File: napari/utils/translations.py
Comment (tlambert03) [reply napari/utils/translations.py]: > Python entry points are "pre-discovered" when the python interpreter is started 

we successfully detect plugins entry points that are installed via the gui, without restarting... so I'm not sure that's the case?  

please have a look at https://github.com/goanpeca/napari/pull/1, which solves all of my concerns... and tell me what you think.

**Comment by tlambert03** [comment]
maybe try it out here?  We routinely discovery plugins installed after napari is started via their entry points...

**Comment by tlambert03** [comment]
it's possible that what you've observed is a limitation of setuptools  and `pkg_resources` itself?  I'm unaware of that limitation

**Comment by goanpeca** [comment]
>it's possible that what you've observed is a limitation of setuptools and pkg_resources itself? I'm unaware of that limitation

It may be the case, cause the previous code for sure was not finding new entrypoints after startup.

**Comment by tlambert03** [comment]
check out our plugin installer: https://github.com/napari/napari/pull/1357

it can detect things installed after runtime.  (But of course, let me know if it doesn't work for you with my PR to your branch)

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py:26]: Fixed. I forgot to add a method. See bleow.

File: napari/utils/translations.py
Comment (goanpeca) [inline napari/utils/translations.py:150]: @tlambert03 this was the reason for the display name func. This will be used when generating the preferences dialog.

File: MANIFEST.in
Comment (tlambert03) [inline MANIFEST.in]: do we still need this with the new test fixture?

File: napari/utils/translations.py
Comment (tlambert03) [inline napari/utils/translations.py:74]: how would this end up getting installed in typical usage?  Is this something that all language packs will necessarily have in their `install_requires`?  Do we need to add an extra?

Comment (goanpeca) [reply napari/utils/translations.py:74]: >Is this something that all language packs will necessarily have in their install_requires?

This was my suggestion, but the language packs template will be generated automatically on the `language-packs` repo, so we will take care of that anyway.

File: MANIFEST.in
Comment (goanpeca) [reply MANIFEST.in]: Nope we do not! good catch!

**Comment by goanpeca** [comment]
Thanks for the suggestions @tlambert03 and @sofroniewn, I think this should be ready now.

File: napari/utils/_tests/test_translations.py
Comment (tlambert03) [inline napari/utils/_tests/test_translations.py]: can you point me to the failure on CI?
If it's just ubuntu and 3.7, can we skip only that rather than xfailing everything?

**Review by tlambert03 — APPROVED** [review]
looking good @goanpeca

I just tried pip installing at runtime and it seems to work... was that your experience too?

```python
In [1]: from napari.utils import translations

In [2]: translations.get_language_packs()
Out[2]: {'en': {'displayName': 'English', 'nativeName': 'English'}}

In [3]: !pip install /Users/talley/Desktop/napari-language-pack-es_CO
Processing /Users/talley/Desktop/napari-language-pack-es_CO
Building wheels for collected packages: napari-language-pack-es-CO
  Building wheel for napari-language-pack-es-CO (setup.py) ... done
  Created wheel for napari-language-pack-es-CO: filename=napari_language_pack_es_CO-0.1.0-py2.py3-none-any.whl size=2233 sha256=51f4d987575cd1af7ebf60253ac3bbcde58ca22047a9f679a04ff5ed49e90347
  Stored in directory: /Users/talley/Library/Caches/pip/wheels/72/f4/fb/48e30679e5acdaf33b1599b6f406a0d8e5ebb33a1d6b0f32fc
Successfully built napari-language-pack-es-CO
Installing collected packages: napari-language-pack-es-CO
Successfully installed napari-language-pack-es-CO-0.1.0

In [4]: translations.get_language_packs()
Out[4]:
{'en': {'displayName': 'English', 'nativeName': 'English'},
 'es_CO': {'displayName': 'Spanish (colombia)',
  'nativeName': 'Español (colombia)'}}
```

File: napari/utils/_tests/test_translations.py
Comment (goanpeca) [reply napari/utils/_tests/test_translations.py]: Fixed @tlambert03 !

**Comment by goanpeca** [comment]
@tlambert03 added the skips to ubuntu and 3.7. Thanks again for the review!

**Comment by tlambert03** [comment]
this looks great!  thanks for your patience :)

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py]:
```suggestion
    locale : str
```

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py]:
```suggestion
    display_locale : str, optional
```

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py]:
```suggestion
    locale : str
```

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py]:
```suggestion
    display_locale : str, optional
```

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py]:
```suggestion
    def _np(self, msgctxt: str, msgid: str, msgid_plural: str, n: str) -> str:
```

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py]:
```suggestion
        return self.npgettext(msgctxt, msgid, msgid_plural, n)
```

File: napari/utils/translations.py
Comment (sofroniewn) [inline napari/utils/translations.py:196]:
```suggestion
                warnings.warn(f"Requested locale not available: {locale}")
```

Comment (sofroniewn) [reply napari/utils/translations.py:196]: not sure if this needs an `f`. I won't let it hold up merge though, as it is easy to fix on your next localization PR

**Review by sofroniewn — COMMENTED** [review]
Ok, all changes look great now. I may have found one more typo, but it can be fixed after merge. Great to see this begin to take shape!!

**Pull request merged**
